### ISSUE TITLE
String + in output declaration leads to "Not such variable."

### ISSUE THREAD

**Issue by yesimon** [nextflow-io/nextflow#42]
This might not be an bug, but here's a simple nf script to demonstrate. Replacing `val(fastq.baseName + '.out')` with `val('abcd' + '.out')` works.

```
#!/usr/bin/env nextflow
seqs = file('seqs/a.fastq')

process proc1 {
  input:
  file fastq from seqs

  output:
  val(fastq.baseName + '.out') into proc1

  """
  echo 1
  """
}
```

```
N E X T F L O W  ~  version 0.13.4-SNAPSHOT
Launching strcat.nf
ERROR ~ Not such variable: fastq

 -- Check script 'strcat.nf' at line: 9 or see '.nextflow.log' file for more details
```

.nextflow.log:
https://gist.github.com/yesimon/fab422b9442ff83fd0c8

**Comment by pditommaso** [comment]
No, variable value names are not supported.

**Comment by yesimon** [comment]
So this is strange, `val(fastq.baseName)` also works. Is that an accident?

**Comment by pditommaso** [comment]
I think so. I will investigate that. BTW do you have any feedback on #41 ?

**Comment by pditommaso** [comment]
I'm still a bit confused what you were trying to do here. The `val` in the `output` section it is supposed to be used to emit a value, while in the example it seems you were trying to capture a file. 

Could you please elaborate a bit more your use case? 

Thanks, p

**Comment by yesimon** [comment]
Yes the example is just an nf file that can reproduce the bug. The use case is similar to the example in the other bugreport I had submitted.

The example is not so bad. It was intended to be used in a `set val(fastq.baseName), file(fastq.baseName + '.step1')` to save the file with a dynamic filename in an indicated filename instead of having a filename like 'result'. I know that this is not idiomatic nextflow but I'm not really used to the workflow of having to look inside the workdir for each different input file like `work/4a/4a.../result`, and somehow gathering that to my own other directory for future analysis. Is that how it's supposed to work? I know you can combine into one downstream process into a single folder but then you would need the extra `val()` data to label each file which is the same problem.

**Comment by pditommaso** [comment]
It makes sense. If you want to a dynamic file name you need to wrap the expression into a closure. (that's a bit verbose, in a future version should be easier) 
http://www.nextflow.io/docs/latest/process.html#dynamic-output-file-names

This works only for `file` entries. It won't work for the `val` object. However you use case makes sense, I will try to extend the syntax to handle it.

**Comment by yesimon** [comment]
Thanks, right now as a workaround I've been using the `script` block to create a new variable that can be used in `output`. I'm not sure if this is meant to work.

**Comment by pditommaso** [comment]
Yes. `val` is meant to capture a value declared in the execution context. But actually makes sense to use it to a emit a literal value. 

Also I'm thinking you may be interested to the [tag](http://www.nextflow.io/docs/latest/process.html#tag) directive that allows you to label process executions with a custom defined indentifier.

**Comment by pditommaso** [comment]
@yesimon I've just released version `0.15.0` that solves this problem. See #59

**Comment by yesimon** [comment]
Thanks!
